Docker organize

Procdox opened the pull request:
This branch organizes the code into separate sub-projects, and implements docker to contain each sub project as its own process. 
This allows any machine supporting docker, to support any technologies we need in the future, or now, without a complex install process, or different install processes for different versions. 
I think it will also make navigating the project easier, as code for things like migrations, will not be sitting next to the primary flask app.
Lastly, it allows for ease of deployment. Once the project is finished, a dockerized code base will be easy to deploy to any service supporting docker. Such as AWS, gcloud, or in house.

Commit:
Docker Construct

This is a fairly large overhaul, but gives a working example of a base docker setup for flask and mongo. the migration management service will be changed to be more accessible in the future (while hopefully keeping node contained)

Commit:
Volumes

Manage service and mongo both use volumes now, this allows persistence for mongo, and ease of migration editing form manage.

Commit:
readme updates

describing migration methods and key location

Commit:
cleanup

Commit:
shared constants, workdir change, and readme typos

condensed constants.json to a single file, and created support for shared resources.
Apps now operate in /usr/src/app instead of odd root directories.
Readme had various clarifications

Commit:
UNSAFE

Mid-Dev changes to remove now redundant pipenv.
Support for hot loading main flask app changes to container

GerryFudd commented on WEB/controllers/router.py:
```diff
@@ -5,7 +5,7 @@ def add_routes(app):
   # TODO: remove this endpoint before this app goes into production
   @app.route('/')
   def hello_world():
-      return 'Hello, World!'
```
LOL, why?

Procdox replied:
I am currently testing hot loading files through flask and docker, to ease development. This was the most apparent way to test hot loading.

GerryFudd replied:
Makes sense :)

GerryFudd commented on MANAGE/migrate-mongo-config.js:
```diff
@@ -1,6 +1,6 @@
 const {database_name} = require('./constants');
 
-const mongoHost = process.env.MONGO_HOST || 'localhost';
```
What prompted you to change the hostname from localhost to mongo?

Procdox replied:
Within docker, referencing a service name is akin to referencing it on a network. Mongo is no longer running on the localhost, but on a networked container. This points to that container.

GerryFudd commented on MANAGE/package.json:
```diff
@@ -6,9 +6,6 @@
   "scripts": {
     "create": "migrate-mongo create",
     "down": "migrate-mongo down",
-    "postinstall": "pipenv install --skip-lock",
```
Is it typically better practice for python apps to avoid tools like pipenv?

It seems like Docker does the same sorts of things.  Am I correct in understanding that it's a proper replacement?

Procdox replied:
that line should be removed, it was there from the pre-docker branch that relied on pipenv. In fact, the node.js manage sub-project does not directly interact with flask web sub-project

Procdox replied:
Docker is replacing pipenv, yes. The intial build of a docker container creates a sandboxed immutable python application. This means that the image must be rebuilt if any dependencies are added. But documentation of all dependencies is implicit in the development process.

GerryFudd replied:
This is more of a general question about dependency management tools for python.  Is it a more common practice to manage dependencies with a docker task rather than with a separate tool like pipenv?

Procdox replied:
I don’t honestly know, my experience with Python is mostly personal. I found early on, that docker gave me more consistency,  made version management easier, and made removing dependencies simple.

Procdox replied:
pipenv does the double edged sword of not redownloading all dependencies whenever one is added. Though this can lead to a bloated dependency folder if improperly managed.

GerryFudd reviewed (commented):
I will have to try to run this once I'm not at work, but I don't see anything objectionable.

I just have a couple of questions for my own education.

OompahLoompah commented on WEB/Dockerfile:
```diff
@@ -0,0 +1,10 @@
+
+RUN pip3 install -r requirements.txt
+
+CMD ["flask", "run", "--host=0.0.0.0"]
```
Is there a specific reason to use 0.0.0.0 for the host string? I feel kind of twitchy about this since it exposes the application to the outside world by default which seems very wrong for dev environments.

Procdox replied:
It is not exposing it externally, it is exposing it to docker’s private network. Docker connects all services withing a virtual network. However, it is forwarding port 5000 from the virtual network to the localhost. Which is done by the port configuration option in docker-compose.yml

OompahLoompah replied:
0.0.0.0 is shorthand for any IP address resulting in the running server being exposed on all of your machine's IP addresses. I can access the app from other computers on my local network after running `docker-compose up`.

Procdox replied:
Hmm, can you still access it after commenting the port 5000:5000 option?

Procdox replied:
I believe if you change it to port 5000:127.0.0.1:5000 or 127.0.0.1:5000:5000 it should restrict access to localhost

OompahLoompah replied:
It needs to be `127.0.0.1:5000:5000` but if we can set up a build process so that when built locally the port is restricted to `127.0.0.1:5000:5000` and when built on the master branch or on some CI/CD system it's set to `5000:5000` I'd be happy. For now setting it statically to `127.0.0.1:5000:5000` would be fine since we're definitely not putting anything into production soon.

OompahLoompah replied:
FWIW I'd be more than happy to set up that build process tonight.

Procdox replied:
If you want to work on that sure, it’d go well with integrating hot loading, which certainly should not be in the release build. If you want to change that line rn feel free. I’m not able to do it from work.

OompahLoompah replied:
Ok, I've committed the change. I should get back to work myself. 😅

OompahLoompah commented on WEB/app.py:
```diff
@@ -17,4 +17,4 @@ def create_app():
   app = create_app()
-  app.run()
\ No newline at end of file
+  app.run(host="0.0.0.0")
```
Same question here. Is there a specific reason to use 0.0.0.0 for the host string?

Procdox replied:
That is redundant, in case flask is not launched from command line.

OompahLoompah replied:
Coolness. Thanks for the clarification!

OompahLoompah committed:
Restrict app access to localhost

The WEB container listens for connection attempts to it on any IP address
available, including any public IPs. In order to prevent accidental exposure of dev
environmnets to the outside world we need to restrict the container to
only forward connections originating from localhost for now. This will
need to be changed before putting the app into production.

OompahLoompah commented:
If we are doing things properly we should probably get a third party to approve since I've committed changes to this branch.

Commit:
docker-compose documentation and cleanup

Commit:
Dockerfile cache use

Organized Dockerfiles to better use the build cache. Dependencies will only be rebuilt when nessecary.

Commit:
dynamic updates revisited

dynamic updates feels hacked together and defeats the purpose of sandboxing. Via caching and documentation, I think rebuilding docker images should be much less painless.

OompahLoompah commented on docker-compose.yml:
```diff
@@ -0,0 +1,62 @@
+
+
+  mongo: #defines the mongoDB enclosure
+    image: mongo:latest
```
Ah, I missed this in the first go-around. We should set this to a specific version rather than using the `latest` tag. If mongo releases some new fancy version that has breaking changes our builds would begin to mysteriously change. Best practice is to always specify the exact version tag you want to use.

Procdox replied:
We should probably do the same for each image, npm package, and python dependency.

Commit:
Explicit versioning

Procedurally-Unorthodox commented:
@OompahLoompah and myself discussed using extends to create a separate docker-compose build for both an administrative deployment and production deployment. Otherwise, the docker containment is where it needed to be.
